src/main/scala/zio/ftp/package.scala

```diff
@@ -0,0 +1,140 @@
+
+      /**
+       * Check existence of a file. If the file doesn't exist, the operation will succeed by returning an empty ftp resource
+       * If the operation failed an error will be emitted
+       *
+       * @param path absolute path of the file
+       */
+      def stat(path: String): ZIO[Blocking, IOException, Option[FtpResource]]
+
+      /**
+       * Read a file by using stream. If the operation failed, an error will be emitted
+       *
+       * @param path absolute path of a file
+       * @param chunkSize default chunk size is 2048 bytes
+       */
+      def readFile(path: String, chunkSize: Int = 2048): ZStreamChunk[Blocking, IOException, Byte]
+
+      /**
+       * Delete a file on a server. If the operation failed, an error will be emitted
+       *
+       * @param path absolute path of the file
+       */
+      def rm(path: String): ZIO[Blocking, IOException, Unit]
+
+      /**
+       * Delete a directory. If the operation failed, an error will be emitted
+       *
+       * @param path absolute path of the directory
+       * @return
+       */
+      def rmdir(path: String): ZIO[Blocking, IOException, Unit]
+
+      /**
+       * Create a directory. If the operation failed, an error will be emitted
+       *
+       * @param path absolute path of the directory
+       */
+      def mkdir(path: String): ZIO[Blocking, IOException, Unit]
+
+      /**
+       * List of files / directories. If the operation failed, an error will be emitted
+       *
+       * @param path absolute path of the directory
+       */
+      def ls(path: String): ZStream[Blocking, IOException, FtpResource]
+
+      /**
+       * List of files from a base directory recursively. If the operation failed, an error will be emitted
+       *
+       * @param path absolute path of the directory
+       */
+      def lsDescendant(path: String): ZStream[Blocking, IOException, FtpResource]
+
+      /**
+       * Save a data stream. If the operation failed, an error will be emitted
+       *
+       * @param path absolute path of file to store
+       * @param source data stream to store
+       * @tparam R environment of the specified stream source, required to extend Blocking
+       */
+      def upload[R <: Blocking](path: String, source: ZStreamChunk[R, Throwable, Byte]): ZIO[R, IOException, Unit]
+    }
+  }
+
+  def secure(settings: SecureFtpSettings): ZLayer[Blocking, ConnectionError, Ftp[SecureFtp.Client]] =
+    ZLayer.fromManaged(SecureFtp.connect(settings))
+
+  def unsecure(settings: UnsecureFtpSettings): ZLayer[Blocking, ConnectionError, Ftp[UnsecureFtp.Client]] =
+    ZLayer.fromManaged(UnsecureFtp.connect(settings))
+
+  def test(path: ZPath): Layer[Any, Ftp[Unit]] =
+    ZLayer.succeed(TestFtp.create(path))
+
+  def execute[A: Tagged, T](f: A => T): ZIO[Ftp[A] with Blocking, IOException, T] =
+    ZIO.accessM(_.get.execute(f))
+
+  def stat[A: Tagged](path: String): ZIO[Ftp[A] with Blocking, IOException, Option[FtpResource]] =
```

Comment:
@jdegoes 

since i defined a type constructor
```scala
type Ftp[A] = Has[Ftp.Service[A]]
```
 i need to propagate the constructor type,  in all the alias methods, i currently need to provide the Client type parameter

```scala
object App {
  val settings: SecureFtpSettings = SecureFtpSettings("127.0.0.1", port = 2222, FtpCredentials("foo", "foo"))
  val ftp: ZLayer[Any, ConnectionError, Ftp[SecureFtp.Client]] = Blocking.live >>> secure(settings)

  ls[SecureFtp.Client]("test").provideCustomLayer(ftp ++ Blocking.live)
}
```

Since im only using this type parameter for `def execute[A: Tagged, T](f: A => T): ZIO[Ftp[A] with Blocking, IOException, T]` how can we improve this ?

Reply:
Here's what I recommend:

```scala
trait FtpAccessors[+A] {
  def stat(path: String): ZIO[Has[Ftp.Service[A]], ...]
}
```

Then you can make two type aliases:

```scala
type Ftp = Has[Ftp.Service[Ftp.Client]]
type Sftp = Has[Ftp.Service[SecureFtp.Client]]
```

And two companion objects:

```scala
object Ftp extends FtpAccessors[Ftp.Client]
object Sftp extends FtpAccessors[SecureFtp.Client]
```

Then it will be no longer necessary to specify that type parameter.